### ISSUE TITLE
If make a variable on `async with` and use that variable outside of `async with`, it will be occured error.

### ISSUE THREAD

**Issue by tasuren** [microsoft/pyright#3706]
**Describe the bug**
If make a variable on `async with` and use that variable outside of `async with`, it will be occured `reportUnboundVariable`.

**To Reproduce**
Make a variable on `async with`.
And use that variable outside of `async with`.
It will be occured the following error on `soup.find_all`.
```python
"<VariableName>" is possibly unbound
```

**Expected behavior**
No error.

**Screenshots or Code**
```python
async with self.bot.session.get("http://jin115.com") as r:
  soup = BeautifulSoup(await r.text(), "lxml")
for article_soup in soup.find_all("div", class_="index_article_body"):
  ...
```
It will be occured error on `soup.find_all`.

**VS Code extension or command-line**
![image](https://user-images.githubusercontent.com/45121209/179120256-89560165-b6c1-4e5f-8b88-6db791d6d313.jpeg)

**Additional context**
This error is happen on my github codespace.
But this error is not heppen on vscode on my pc.

**Comment by erictraut** [comment]
Please provide a self-contained minimal code sample that exhibits the problem. The code sample you provided above references symbols that are presumably imported from other libraries. If your minimal sample relies on libraries, please include the versions of those libraries that you have installed and specify whether you have any type stubs for those libraries.

**Comment by BlueGlassBlock** [comment]
@erictraut I think this is probably due to ec664a1abb06c8a18ddf19e543708691e9c6d85c, which changed the behavior of handling context managers.

For example, this simple snippet will emit type issues:
```py
from contextlib import contextmanager


@contextmanager
def wrap():
    print("enter")
    yield
    print("exit")


with wrap():
    q = 5

print(q) # "q" is possibly unbound
```

Which I think is undesirable, because the `wrap` function **doesn't** mask the error happened inside it by returning a True value, which means the "q" will always be bound.

**Comment by BlueGlassBlock** [comment]
Further digesting the code let me realized that pyright will treat the function as "maybe masking the error" because `_GeneratorContextManager` wrapped `__exit__` only annotated the return type as `bool | None`, which could explain pyright's behavior.

**Comment by erictraut** [comment]
@BlueGlassBlock, I'm not able to repro any errors for that code sample. Can you verify that you are seeing an error? There should be no error in this case because a context manager created with the `@contextmanager` decorator has an `__exit__` method that is annotated with a return type of `bool | None` which indicates to a type checker that the context manager will not swallow exceptions.

**Comment by BlueGlassBlock** [comment]
@erictraut OK, Digesting the commit history again let me realized that it's the problem is fixed in pyright `1.1.260`, which is currently shipped with Pylance `2022.7.31`, a preview version. I think this could be considered as fixed.

**Comment by erictraut** [comment]
There have been no real changes in this behavior for about a year. The recent commit added support specifically for an `__exit__` return type of `Literal[False]`, so your sample above wouldn't be affected by the recent change. I don't think your conclusion that "the problem is fixed in `1.1.260`" is correct. 

You provided a sample above that seemed to imply that you were seeing an error `"q" is possibly unbound`. Can you confirm that you were seeing this error? If so, what version of pyright or pylance were you using at the time?

**Comment by BlueGlassBlock** [comment]
> You provided a sample above that seemed to imply that you were seeing an error `"q" is possibly unbound`. Can you confirm that you were seeing this error? If so, what version of pyright or pylance were you using at the time?

@erictraut This incorrect behavior could be reproduced with Pylance `2022.7.30`, however Pylance `2022.7.31` functions well.
![image](https://user-images.githubusercontent.com/50769752/179139465-e2ddfd6d-547b-49f0-b0e4-1f86829abe0f.png)
![image](https://user-images.githubusercontent.com/50769752/179139850-a2b0b210-6675-4acd-a851-3e2ca8c5e996.png)

**Comment by erictraut** [comment]
OK, thanks for the additional information. There was a regression that I introduced with a recent change to this code in [this commit](https://github.com/microsoft/pyright/commit/5763cf1abbb06e3b31422b74340afa11ab59d7b2), but I found and fixed the problem one day later in [this commit](https://github.com/microsoft/pyright/commit/ec664a1abb06c8a18ddf19e543708691e9c6d85c). The regression was never released in a published version of pyright. It sounds like pylance `2022.7.30` picked up that bad commit. We make an effort to synchronize pylance and pyright releases (typically on Wed mornings) to prevent differences like this, but that sync process has broken down a bit lately.

In any event, I'm glad to hear that the problem is fixed for you now. I'll wait to hear back from the OP before closing this issue because I'm not convinced that it's the same issue.

**Comment by erictraut** [comment]
@tasuren, since I haven't heard back from you, I'm going to assume for now that the issue has been fixed. If you think that your issue was different from the one I discussed with @BlueGlassBlock above, please feel free to reply, and I'll reopen the issue.

**Comment by tasuren** [comment]
Sorry for the late reply.
I too think this issue has been resolved.
Thank you for your cooperation.

**Comment by asleep-cult** [comment]
I was under the impression that pyright was performing some kind of code flow analysis within the block and the context manager to help determine which code paths could raise an exception (and thus which variables could be unbound and which code paths are unreachable). I actually ended up writing quite a bit of code under this impression so I think there is still something to be desired here.
```py
from contextlib import contextmanager

@contextmanager
def wrap():
    try:
        yield
    except AssertionError:
        pass

with wrap():
    q = 5
    assert False

print(q)
 ```
pyright currently marks anything past the assertion statement as unreachable even though the context manager swallows AssertionErrors. I assume this is an bug in itself but I am left wondering if there could be some kind of special return type for `__exit__` enabling this kind of analysis.

**Comment by erictraut** [comment]
If you implement your own context manager class and declare the return type of `__exit__` to be `None`, type checkers like pyright and mypy will interpret that to mean that your context manager swallows all exceptions. However, in your example above, you're using the `@contextmanager` decorator rather than providing your own context manager class. The `@contextmanager` decorator uses a context manager class called `_GeneratorContextManager`, and its `__exit__` method is declared to return a `bool | None`. This tells type checkers that they should assume that the context manager does not swallow exceptions.
